core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -187,7 +186,7 @@ protected NodesAndVersions buildAllocationIdBasedNodes(ShardRouting shard, boole
             }
 
             if (nodeShardState.storeException() == null) {
-                if (allocationId == null && nodeShardState.version() != -1) {
+                if (allocationId == null && nodeShardState.legacyVersion() != -1) {
```

Comment:
not sure I follow - the snapshot should have been made in 2.0 so we should have an allocation id? also, remind me - does the restore procedure re uses the old shard state or creates a new one?